File: src/main/java/destiny/core/chinese/SimpleBranch.java
/**
 * @author smallufo
 * @date 2002/8/12
 * @time 上午 02:57:12
 */
package destiny.core.chinese;

import org.jetbrains.annotations.NotNull;

/** 實作 五行 getFiveElement() 以及 陰陽 getYinYang() 以及取得地支順序 getIndex() 的地支 */
public enum SimpleBranch implements IBranch<SimpleBranch>, IFiveElement, IYinYang {
  子(Branch.子),
  丑(Branch.丑),
  寅(Branch.寅),
  卯(Branch.卯),
  辰(Branch.辰),
  巳(Branch.巳),
  午(Branch.午),
  未(Branch.未),
  申(Branch.申),
  酉(Branch.酉),
  戌(Branch.戌),
  亥(Branch.亥);
  
  
  @NotNull
  private final Branch eb;

  SimpleBranch(@NotNull Branch eb) {
    this.eb = eb;
  }

  public static SimpleBranch get(Branch b) {
    return SimpleBranch.valueOf(b.name());
  }
  
  @NotNull
  public FiveElement getFiveElement() {
    return getFiveElement(eb);
  }
  
  /**
   * 是否還要再做一個 static FiveElement getFiveElement(SimpleEarthlyBranches eb) ... ??
   * Java 5 中的 enum 無法被繼承 , 真是麻煩... 
   */
  @NotNull
  public static FiveElement getFiveElement(@NotNull Branch eb) {
    switch (eb) {
      case 亥:
      case 子:
        return FiveElement.水;
      case 丑:
      case 辰:
      case 未:
      case 戌:
        return FiveElement.土;
      case 寅:
      case 卯:
        return FiveElement.木;
      case 巳:
      case 午:
        return FiveElement.火;
      case 申:
      case 酉:
        return FiveElement.金;
    }
    throw new AssertionError("getFiveElement() of " + eb);
  }

  @Override
  public boolean getBooleanValue() {
    return getBooleanValue(this.getBranch());
  }

  public final static boolean getBooleanValue(Branch branch) {
    return (Branch.Companion.getIndex(branch) % 2 == 0);
  }


  @Override
  public int getAheadOf(SimpleBranch other) {
    return eb.getAheadOf(other.eb);
  }

  @Override
  public Branch getBranch() {
    return eb;
  }
}

File: src/main/java/destiny/core/chinese/SimpleBranch.kt
/**
 * @author smallufo
 * @date 2002/8/12
 * @time 上午 02:57:12
 */
package destiny.core.chinese

/** 實作 五行 [IFiveElement] 以及 陰陽 [IYinYang] 以及取得地支順序 getIndex() 的地支  */
enum class SimpleBranch(override val branch: Branch) : IBranch<SimpleBranch>, IFiveElement, IYinYang {
  子(Branch.子),
  丑(Branch.丑),
  寅(Branch.寅),
  卯(Branch.卯),
  辰(Branch.辰),
  巳(Branch.巳),
  午(Branch.午),
  未(Branch.未),
  申(Branch.申),
  酉(Branch.酉),
  戌(Branch.戌),
  亥(Branch.亥);


  override val fiveElement: FiveElement
    get() = Companion.getFiveElement(branch)


  override val booleanValue: Boolean
    get() = Companion.getBooleanValue(this.branch)


  override fun getAheadOf(other: SimpleBranch): Int {
    return branch.getAheadOf(other.branch)
  }

  companion object {

    operator fun get(b: Branch): SimpleBranch {
      return SimpleBranch.valueOf(b.name)
    }

    /**
     * 是否還要再做一個 static FiveElement getFiveElement(SimpleEarthlyBranches eb) ... ??
     * Java 5 中的 enum 無法被繼承 , 真是麻煩...
     */
    fun getFiveElement(b: Branch): FiveElement {
      return when (b) {
        Branch.亥, Branch.子 -> FiveElement.水
        Branch.丑, Branch.辰, Branch.未, Branch.戌 -> FiveElement.土
        Branch.寅, Branch.卯 -> FiveElement.木
        Branch.巳, Branch.午 -> FiveElement.火
        Branch.申, Branch.酉 -> FiveElement.金
      }
    }

    fun getBooleanValue(branch: Branch): Boolean {
      return Branch.getIndex(branch) % 2 == 0
    }
  }
}
